slapo/utils/common.py

```diff
@@ -0,0 +1,55 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Common utilities used in schedule."""
+
+
+def get_hooks(mod):
+    """Get the hooks of a module.
+
+    Parameters
+    ----------
+    mod : torch.nn.Module
+        The module.
+
+    Returns
+    -------
+    dict
+        A dictionary of hooks.
+    """
+    hooks = {"fwd_pre": [], "fwd_post": [], "bwd_post": []}
+    for hook in mod._forward_hooks.values():
+        hooks["fwd_post"].append(hook)
+
+    for hook in mod._forward_pre_hooks.values():
+        hooks["fwd_pre"].append(hook)
+
+    for hook in mod._backward_hooks.values():
+        hooks["bwd_post"].append(hook)
+
+    return hooks
+
+
+def transfer_hooks(old_mod, new_mod, hook_types=None):
+    """Transfer the hooks from old_mod to new_mod.
+
+    Parameters
+    ----------
+    old_mod : torch.nn.Module
+        The old module.
+    new_mod : torch.nn.Module
+        The new module.
+    hook_types : Optional[List[str]]
+        The types of hooks to transfer. If None, transfer all hooks.
+    """
+    HOOK_TYPE_TO_ATTR = {
+        "fwd_pre": ("_forward_pre_hooks", "register_forward_pre_hook"),
+        "fwd_post": ("_forward_hooks", "register_forward_hook"),
+        "bwd_post": ("_backward_hooks", "register_backward_hook"),
```

Comment:
why don't we also include bwd_pre?

Reply:
I don't think there's a backward pre-hook in PyTorch?

Reply:
just found this: https://github.com/pytorch/pytorch/blob/b90496eef5665bc39828f6c1c522f399bcc62f3f/torch/nn/modules/module.py#L92. and it seems it does not appear in 1.12